pkg/inventory/db.go

```diff
@@ -200,25 +170,131 @@ func (db *DB) GetResources(ctx context.Context) <-chan []resourceapi.Device {
+func (db *DB) discoverNetworkInterfaces(pciDevices []resourceapi.Device) []resourceapi.Device {
+	links, err := netlink.LinkList()
+	if err != nil {
+		klog.Errorf("Could not list network interfaces: %v", err)
+		return pciDevices
+	}
+
+	pciDeviceMap := make(map[string]*resourceapi.Device)
+	for i := range pciDevices {
+		pciDeviceMap[pciDevices[i].Name] = &pciDevices[i]
+	}
+
+	otherDevices := []resourceapi.Device{}
+
+	for _, link := range links {
+		ifName := link.Attrs().Name
+		if ignoredInterfaceNames.Has(ifName) {
+			klog.V(4).Infof("Network Interface %s is in the list of ignored interfaces, excluding it from discovery", ifName)
+			continue
+		}
+
+		// skip loopback interfaces
+		if link.Attrs().Flags&net.FlagLoopback != 0 {
+			klog.V(4).Infof("Network Interface %s is a loopback interface, excluding it from discovery", ifName)
+			continue
+		}
+
+		pciAddr, err := bdfAddress(ifName, sysnetPath)
+		if err == nil {
+			// It's a PCI device.
+
+			normalizedAddress := names.NormalizePCIAddress(pciAddr.String())
+			var exists bool
+			device, exists := pciDeviceMap[normalizedAddress]
+			if !exists {
+				// We don't expect this to happen.
+				klog.Errorf("Network interface %s has PCI address %q, but it was not found in initial PCI scan.", ifName, pciAddr)
+				continue
+			}
+			addLinkAttributes(device, link)
+		} else {
+			// Not a PCI device.
+
+			if !isVirtual(ifName, sysnetPath) {
+				// If we failed to identify the PCI address of the network
+				// interface and the network interface is also not a virtual
+				// device, use a best-effort strategy where the network
+				// interface is assumed to be virtual.
+				klog.Warningf("PCI address not found for non-virtual interface %s, proceeding as if it were virtual. Error: %v", ifName, err)
+			}
+			newDevice := &resourceapi.Device{
+				Name:       names.NormalizeInterfaceName(ifName),
+				Attributes: make(map[resourceapi.QualifiedName]resourceapi.DeviceAttribute),
+			}
+			addLinkAttributes(newDevice, link)
+			otherDevices = append(otherDevices, *newDevice)
+		}
+	}
+
+	return append(pciDevices, otherDevices...)
+}
+
+func addLinkAttributes(device *resourceapi.Device, link netlink.Link) {
+	ifName := link.Attrs().Name
+	device.Attributes[apis.AttrInterfaceName] = resourceapi.DeviceAttribute{StringValue: &ifName}
+	device.Attributes[apis.AttrMac] = resourceapi.DeviceAttribute{StringValue: ptr.To(link.Attrs().HardwareAddr.String())}
+	device.Attributes[apis.AttrMTU] = resourceapi.DeviceAttribute{IntValue: ptr.To(int64(link.Attrs().MTU))}
+	device.Attributes[apis.AttrEncapsulation] = resourceapi.DeviceAttribute{StringValue: ptr.To(link.Attrs().EncapType)}
+	device.Attributes[apis.AttrAlias] = resourceapi.DeviceAttribute{StringValue: ptr.To(link.Attrs().Alias)}
+	device.Attributes[apis.AttrState] = resourceapi.DeviceAttribute{StringValue: ptr.To(link.Attrs().OperState.String())}
+	device.Attributes[apis.AttrType] = resourceapi.DeviceAttribute{StringValue: ptr.To(link.Type())}
```

Comment:
I think some of this attributes may not exist, like Alias and I think also MTU, shouldn't we only add them if exist?